File: backend/app/api/v1/contacts.py
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.deps import TenantContext, require_permission
from app.db.session import get_db
from app.schemas.contact import ContactCreate, ContactListResponse, ContactResponse, ContactUpdate
from app.services.contact_service import ContactService, paginate

router = APIRouter(prefix="/tenants/{slug}/contacts", tags=["contacts"])


def _to_response(contact) -> ContactResponse:
    assigned = None
    if contact.assigned_to:
        assigned = {
            "id": contact.assigned_to.id,
            "full_name": contact.assigned_to.full_name,
            "email": contact.assigned_to.email,
        }
    lead_ref = None
    if contact.lead:
        lead_ref = {
            "id": contact.lead.id,
            "first_name": contact.lead.first_name,
            "last_name": contact.lead.last_name,
            "status": contact.lead.status,
        }
    company_ref = None
    if contact.linked_company:
        company_ref = {
            "id": contact.linked_company.id,
            "company_name": contact.linked_company.company_name,
            "company_code": contact.linked_company.company_code,
        }
    return ContactResponse(
        id=contact.id,
        tenant_id=contact.tenant_id,
        lead_id=contact.lead_id,
        company_id=contact.company_id,
        first_name=contact.first_name,
        last_name=contact.last_name,
        email=contact.email,
        phone=contact.phone,
        company=contact.company,
        job_title=contact.job_title,
        notes=contact.notes,
        assigned_to_id=contact.assigned_to_id,
        assigned_to=assigned,
        lead=lead_ref,
        linked_company=company_ref,
        created_by_id=contact.created_by_id,
        created_at=contact.created_at,
        updated_at=contact.updated_at,
    )


@router.get("", response_model=ContactListResponse)
def list_contacts(
    q: str | None = Query(default=None, max_length=200),
    company: str | None = Query(default=None, max_length=255),
    company_id: UUID | None = Query(default=None),
    assigned_to_id: UUID | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    sort_by: str = Query(default="created_at"),
    sort_order: str = Query(default="desc", pattern="^(asc|desc)$"),
    ctx: TenantContext = Depends(require_permission("contact:read")),
    db: Session = Depends(get_db),
) -> ContactListResponse:
    service = ContactService(db)
    contacts, total = service.list_contacts(
        ctx.tenant.id,
        q=q,
        company=company,
        company_id=company_id,
        assigned_to_id=assigned_to_id,
        page=page,
        page_size=page_size,
        sort_by=sort_by,
        sort_order=sort_order,
    )
    meta = paginate(total, page, page_size)
    return ContactListResponse(
        items=[_to_response(contact) for contact in contacts],
        **meta,
    )


@router.post("", response_model=ContactResponse, status_code=201)
def create_contact(
    payload: ContactCreate,
    ctx: TenantContext = Depends(require_permission("contact:write")),
    db: Session = Depends(get_db),
) -> ContactResponse:
    contact = ContactService(db).create_contact(
        ctx.tenant.id, payload, ctx.membership.user_id
    )
    return _to_response(contact)


@router.post("/convert-lead/{lead_id}", response_model=ContactResponse, status_code=201)
def convert_lead_to_contact(
    lead_id: UUID,
    ctx: TenantContext = Depends(require_permission("contact:write")),
    db: Session = Depends(get_db),
) -> ContactResponse:
    contact = ContactService(db).convert_lead(
        ctx.tenant.id, lead_id, ctx.membership.user_id
    )
    return _to_response(contact)


@router.get("/{contact_id}", response_model=ContactResponse)
def get_contact(
    contact_id: UUID,
    ctx: TenantContext = Depends(require_permission("contact:read")),
    db: Session = Depends(get_db),
) -> ContactResponse:
    contact = ContactService(db).get_contact(ctx.tenant.id, contact_id)
    return _to_response(contact)


@router.put("/{contact_id}", response_model=ContactResponse)
def update_contact(
    contact_id: UUID,
    payload: ContactUpdate,
    ctx: TenantContext = Depends(require_permission("contact:write")),
    db: Session = Depends(get_db),
) -> ContactResponse:
    contact = ContactService(db).update_contact(ctx.tenant.id, contact_id, payload)
    return _to_response(contact)


@router.delete("/{contact_id}", status_code=204)
def delete_contact(
    contact_id: UUID,
    ctx: TenantContext = Depends(require_permission("contact:delete")),
    db: Session = Depends(get_db),
) -> None:
    ContactService(db).delete_contact(ctx.tenant.id, contact_id)

File: frontend/lib/api/contacts.ts
import { apiFetch } from "@/lib/api/client";
import type { Contact, ContactFilters, ContactListResponse } from "@/types/api";

export type ContactInput = {
  first_name: string;
  last_name?: string;
  email?: string | null;
  phone?: string | null;
  company?: string | null;
  company_id?: string | null;
  job_title?: string | null;
  notes?: string | null;
  lead_id?: string | null;
  assigned_to_id?: string | null;
};

function buildQuery(filters: ContactFilters): string {
  const params = new URLSearchParams();
  if (filters.q) params.set("q", filters.q);
  if (filters.company) params.set("company", filters.company);
  if (filters.company_id) params.set("company_id", filters.company_id);
  if (filters.assigned_to_id) params.set("assigned_to_id", filters.assigned_to_id);
  if (filters.page) params.set("page", String(filters.page));
  if (filters.page_size) params.set("page_size", String(filters.page_size));
  if (filters.sort_by) params.set("sort_by", filters.sort_by);
  if (filters.sort_order) params.set("sort_order", filters.sort_order);
  const query = params.toString();
  return query ? `?${query}` : "";
}

export async function listContacts(
  slug: string,
  filters: ContactFilters = {},
): Promise<ContactListResponse> {
  return apiFetch<ContactListResponse>(`/tenants/${slug}/contacts${buildQuery(filters)}`);
}

export async function getContact(slug: string, contactId: string): Promise<Contact> {
  return apiFetch<Contact>(`/tenants/${slug}/contacts/${contactId}`);
}

export async function createContact(slug: string, data: ContactInput): Promise<Contact> {
  return apiFetch<Contact>(`/tenants/${slug}/contacts`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function updateContact(
  slug: string,
  contactId: string,
  data: ContactInput,
): Promise<Contact> {
  return apiFetch<Contact>(`/tenants/${slug}/contacts/${contactId}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

export async function deleteContact(slug: string, contactId: string): Promise<void> {
  await apiFetch<void>(`/tenants/${slug}/contacts/${contactId}`, { method: "DELETE" });
}

export async function convertLeadToContact(slug: string, leadId: string): Promise<Contact> {
  return apiFetch<Contact>(`/tenants/${slug}/contacts/convert-lead/${leadId}`, {
    method: "POST",
  });
}

export function formatContactName(contact: Contact): string {
  return [contact.first_name, contact.last_name].filter(Boolean).join(" ");
}

export function formatDate(value: string): string {
  return new Date(value).toLocaleDateString("en-US", {
    year: "numeric",
    month: "short",
    day: "numeric",
  });
}
